### PR TITLE
Structured bindings support for gridtools tuple-likes

### PR DESCRIPTION
**Pull request by anstaf**

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

File: include/gridtools/common/array.hpp [hunk 1]
```diff
@@ -197,3 +184,14 @@ namespace gridtools {
     /** @} */
 
 } // namespace gridtools
+
+namespace std {
+    template <class T, size_t D>
```
Comment (havogt) [inline include/gridtools/common/array.hpp:189]: Did we never use `gridtools::tuple_size` etc.?

Comment (havogt) [reply include/gridtools/common/array.hpp:189]: I see, we did but the abstraction is via tuple_util.

File: include/gridtools/common/tuple.hpp [hunk 2]
```diff
@@ -283,4 +285,29 @@ namespace gridtools {
...
+
+namespace std {
+    template <class... Ts>
+    struct tuple_size<::gridtools::tuple<Ts...>> : integral_constant<size_t, sizeof...(Ts)> {};
+
+    template <size_t I, class... Ts>
+    struct tuple_element<I, ::gridtools::tuple<Ts...>> {
+        using type = gridtools::meta::at_c<::gridtools::tuple<Ts...>, I>;
+    };
+} // namespace std
```
[the remark this answers is not in the document]
Comment (havogt) [reply include/gridtools/common/tuple.hpp:313]: (or maybe newline is actually not part of clang-format?)

File: tests/unit_tests/common/test_tuple.cpp [hunk 3]
```diff
@@ -322,5 +322,12 @@ namespace gridtools {
 
         TEST(nested_empty, regression) { EXPECT_EQ(42, get<0>(get<0>(tuple<tuple<an_empty<42>>>())).value); }
 
+#if __cplusplus >= 201703
+        TEST(structured_bindings, functoinal) {
```
Comment (havogt) [inline tests/unit_tests/common/test_tuple.cpp:326]:
```suggestion
        TEST(structured_bindings, functional) {
```

**Review by havogt — APPROVED** [review]
Only the typo, the rest are comments.
I didn't know you can enable structured bindings for your custom types, but totally makes sense. Cool!

File: include/gridtools/common/tuple.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/common/tuple.hpp]: done

File: tests/unit_tests/common/test_tuple.cpp (hunk 3 above)
Comment (anstaf) [reply tests/unit_tests/common/test_tuple.cpp]: done

File: include/gridtools/common/array.hpp (hunk 1 above)
Comment (anstaf) [reply include/gridtools/common/array.hpp:189]: There was one usage of it. I have replaced it to `tuple_util` stuff.

**Comment by anstaf** [comment]
launch jenkins

**Comment by anstaf** [comment]
launch jenkins

**Pull request merged**